### PR TITLE
release-21.1: kvserver: synchronize replica removal with read-only requests

### LINKED ISSUE #64325: kvserver: reads during replica removal can return empty results
Read requests submitted during replica removal may be evaluated on the removed replica, returning empty results. This was discovered when restoring a backup did not recover all of the data -- an entire range was missing. This was caused by the export request racing with a range merge and replica removal, where the export request was evaluated on the removed (empty) replica, returning an empty result and omitting the range from the backup.

### PR DESCRIPTION
**Pull request by erikgrinaker**
Backport 1/1 commits from #64324.

/cc @cockroachdb/release @cockroachdb/kv 

---

Replica removal did not synchronize with in-flight read-only requests,
which could cause them to be evaluated on a removed (empty) replica,
returning an empty result.

This patch fixes the problem by locking `Replica.readOnlyCmdMu` during
replica removal, thus either waiting for read-only requests to complete
or not evaluating them.

Resolves #64325.

Release note (bug fix): Fixed a race condition where read-only requests
during replica removal (e.g. during range merges or rebalancing) could
be evaluated on the removed replica, returning an empty result.

### CODE REVIEW

File: pkg/kv/kvserver/replica_destroy.go
Comment (tbg) [inline pkg/kv/kvserver/replica_destroy.go:189]: This is because it's already held? Can we `r.readOnlyCmdMu.AssertHeld()` in that case (maybe on master only)?

**Review by tbg — APPROVED** [review]
If any part of this was done manually, please point it out and I'll pore over it in detail once more.

File: pkg/kv/kvserver/replica_destroy.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_destroy.go:189]: No, it's not held, but both callers have already called `destroyStatus.Set()` with it (temporarily) held. As such, the following code paths (including this one) are guaranteed to not conflict with read-only requests since they will either have completed already or will check `destroyStatus` and abort.

**Comment by erikgrinaker** [comment]
>If any part of this was done manually, please point it out and I'll pore over it in detail once more.

Just the inclusion of #64378 as well, I'll wait for that to be reviewed again and merged before merging the backports.

File: pkg/kv/kvserver/replica_destroy.go
Comment (nvanbenschoten) [reply pkg/kv/kvserver/replica_destroy.go:189]: The point about using `r.readOnlyCmdMu.AssertHeld()` to assert proper locking is still a good one. Did you give any thought to pulling accesses of `r.mu.destroyStatus` into methods on the replica and asserting proper handling of `r.mu` and `r.readOnlyCmdMu`?

> maybe on master only

That shouldn't be needed. `AssertHeld()`/`AssertRHeld()` should be no-ops on prod builds. The assertions are only armed in race builds.

**Review by nvb — APPROVED** [review]
LGTM

File: pkg/kv/kvserver/replica_destroy.go
Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_destroy.go:189]: > Did you give any thought to pulling accesses of r.mu.destroyStatus into methods on the replica and asserting proper handling of `r.mu` and `r.readOnlyCmdMu`?

That's a good idea. I'm going to merge this and the other backports, so we have a fix ready to go, and then submit a follow-up PR to clean up the lock handling a bit. I'd also like to write a quick test for the write path in #46329 first, to see if it has a similar problem.

Comment (erikgrinaker) [reply pkg/kv/kvserver/replica_destroy.go:189]: #64459

**Pull request merged**
